tests/linearizability/watch.go

```diff
@@ -85,61 +85,94 @@ func validateWatchResponses(t *testing.T, responses [][]watchResponse, expectPro
+	validateGotAtLeastOneProgressNotify(t, responses, expectProgressNotify)
 
-	for _, resp := range responses {
-		if resp.Header.Revision < lastHeadRevision {
-			t.Errorf("Server revision should never decrease, lastHeadRevision: %d, resp.Header.Revision: %d",
-				lastHeadRevision, resp.Header.Revision)
-		}
+	// Validate etcd watch properties
+	validateEventsWithRevisionLowerThanProgressNotifyWereDelivered(t, responses)
+	validateEventRevisionIsContinuousAndNonDecreasing(t, responses)
+	validateWatchEventsAreUnique(t, responses)
+	validateHeaderRevisionGreaterOrEqualToEventRevision(t, responses)
+	validateEventsSharingRevisionWithinSameResponse(t, responses)
+}
 
+func validateGotAtLeastOneProgressNotify(t *testing.T, responses []watchResponse, expectProgressNotify bool) {
+	var gotProgressNotify = false
+	var lastHeadRevision int64 = 1
+	for _, resp := range responses {
 		if resp.IsProgressNotify() && resp.Header.Revision == lastHeadRevision {
 			gotProgressNotify = true
 		}
-		if resp.Header.Revision == lastHeadRevision && len(resp.Events) != 0 {
-			t.Errorf("Got two non-empty responses about same revision")
-		}
-		if len(resp.Events) > 0 && resp.Events[0].Kv.ModRevision <= lastEventModRevision {
-			t.Errorf("Events with same revision should not be split between multiple responses, lastResponseModRevision: %d, firstEventModRevision: %d", lastEventModRevision, resp.Events[0].Kv.ModRevision)
-		}
+		lastHeadRevision = resp.Header.Revision
+	}
+	if gotProgressNotify != expectProgressNotify {
+		t.Errorf("Expected at least one progress notify: %v, got: %v", expectProgressNotify, gotProgressNotify)
+	}
+}
 
+func validateEventsWithRevisionLowerThanProgressNotifyWereDelivered(t *testing.T, responses []watchResponse) {
+	var lastProgressNotifyRevision int64 = 0
+	for _, resp := range responses {
 		for _, event := range resp.Events {
-			rk := revisionKey{key: string(event.Kv.Key), revision: event.Kv.ModRevision}
-			if _, found := uniqueOperations[rk]; found {
-				t.Errorf("BROKE: Within the:wq same revision key can only be modified once. Suspecting duplicate watch event. key: %q, revision: %d", rk.key, rk.revision)
-			}
-			uniqueOperations[rk] = struct{}{}
 			if event.Kv.ModRevision <= lastProgressNotifyRevision {
-				t.Errorf("BROKE: etcd will not send progress notification to a watcher until it has synced all events. So a watcher will never receive an event with a lower `ModRevision` than the last progressNotification's revision, eventRevision: %d, progressNotifyRevision: %d", event.Kv.ModRevision, lastProgressNotifyRevision)
+				t.Errorf("BROKE: Events with revision lower than progress notify were delivered, eventRevision: %d, progressNotifyRevision: %d", event.Kv.ModRevision, lastProgressNotifyRevision)
 			}
+		}
+		if resp.IsProgressNotify() {
+			lastProgressNotifyRevision = resp.Header.Revision
+		}
+	}
+}
+
+func validateEventRevisionIsContinuousAndNonDecreasing(t *testing.T, responses []watchResponse) {
```

Comment:
I raised the comment `Events in the same watch response must have the same revision, otherwise it must be strictly incremental across responses`,  but eventually removed the comment after seeing function `validateEventsSharingRevisionWithinSameResponse`.

Please merge it and `validateEventsSharingRevisionWithinSameResponse` into just one function.  It's clearer to get them included in one function instead of two.

Reply:
I'd like to repeat the removed comment: `Events in the same watch response must have the same revision, otherwise it must be strictly incremental across responses`.  The two functions are actually validating this rule. Please consider to define just one function.